### PR TITLE
Improved PerformanceMonitor

### PR DESCRIPTION
**Pull request by pveentjer**
It includes a lot more flexible monitortask plugin system and it contains a bunch of interesting tasks like packet sampler and invocation occurrence counter that count the number of pending operations aggregated based on occurrence.

A lot of work was done to make sure the PerformanceMonitor/PerformanceLog when enabled, doesn't generate a lot of litter. Most of the litter created is from the MetricsRegistry due to reflection method calls. Unfortunately all returned primitives are wrapped in an object.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/PendingInvocationsMonitorTask.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        }
+    }
+
+    private void render(BufferedWriter writer) throws IOException {
+        writer.write("PendingInvocations[");
+        writeKeyValue(writer, "count", invocationRegistry.size());
+        writer.write(getIndent() + "invocations[");
+        incIndentLevel();
+
+        for (Class op : occurrenceMap.keySet()) {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/PendingInvocationsMonitorTask.java]: it would be nice to sort categories by number of occurencies

**Comment by jerrinot** [comment]
It would be nice to have an ability to monitor reasons why invocations are pending. 

I observed issues with huge numbers (800+) of smart clients. What I see is that most of pending invocations have a response already set (because the target member is local), but they are waiting on backup response.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/renderers/CommaSeparatedKeyValueProbeRenderer.java [hunk 2]
```diff
@@ -32,7 +32,7 @@
     @Override
     public void start() {
         sb.setLength(0);
-        sb.append("{").append("time").append("=").append(System.currentTimeMillis());
+        sb.append("").append("time").append("=").append(System.currentTimeMillis());
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/renderers/CommaSeparatedKeyValueProbeRenderer.java]: This line could need some optimizing :)
`sb.append("time=").append(System.currentTimeMillis());`

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/BuildInfoPlugin.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    private final BuildInfo buildInfo = BuildInfoProvider.getBuildInfo();
+    private final ILogger logger;
+
+    public BuildInfoPlugin(NodeEngineImpl nodeEngine) {
+        this.logger = nodeEngine.getLogger(BuildInfoPlugin.class);
+    }
+
+    @Override
+    public long getPeriodMillis() {
+        return -1;
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/BuildInfoPlugin.java]: This looks like a magic number. I think I'll see this further in the code, but if this means "execute once" it would be nice to have a constant in `PerformanceMonitorPlugin` for this special value.

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/BuildInfoPlugin.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Override
+    public void onStart() {
+        logger.info("Plugin:active, period-seconds:" + getPeriodMillis());
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/BuildInfoPlugin.java]: `period-seconds` in the String vs. `getPeriodMillis()` as method name

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/MetricsPlugin.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    private final MetricsRegistry metricsRegistry;
+    private final int periodSeconds;
+    private final ILogger logger;
+    private final NodeEngineImpl nodeEngine;
+    private ProbeRenderer probeRenderer;
+
+    public MetricsPlugin(NodeEngineImpl nodeEngine) {
+        this.nodeEngine = nodeEngine;
+        this.metricsRegistry = nodeEngine.getMetricsRegistry();
+        this.periodSeconds = nodeEngine.getGroupProperties().getInteger(PERFORMANCE_MONITOR_METRICS_PERIOD_SECONDS);
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/MetricsPlugin.java]: You could use `getMillis()` instead of `getInteger()` here, so you don't have to convert the value later.

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/OverloadedConnectionsPlugin.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+import java.util.Set;
+
+import static com.hazelcast.instance.GroupProperty.PERFORMANCE_MONITOR_OVERLOADED_CONNECTIONS_PERIOD_SECONDS;
+import static com.hazelcast.instance.GroupProperty.PERFORMANCE_MONITOR_OVERLOADED_CONNECTIONS_SAMPLES;
+import static com.hazelcast.instance.GroupProperty.PERFORMANCE_MONITOR_OVERLOADED_CONNECTIONS_THRESHOLD;
+import static java.lang.Math.min;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * The PacketSampler is a debug aid that periodically scans the Connection.writeQueues and samples the content.
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/OverloadedConnectionsPlugin.java]: `PacketSampler` in JavaDoc vs. `OverloadedConnectionsPlugin` as class name

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/OverloadedConnectionsPlugin.java [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    private Set<TcpIpConnection> getTcpIpConnections() {
+        ConnectionManager connectionManager = nodeEngine.getNode().getConnectionManager();
+        if (connectionManager instanceof TcpIpConnectionManager) {
+            return ((TcpIpConnectionManager) connectionManager).getActiveConnections();
+        } else {
+            return Collections.emptySet();
+        }
+    }
+
+    private void scan(BufferedWriter w, TcpIpConnection connection, boolean priority) throws IOException {
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/OverloadedConnectionsPlugin.java]: I would prefer `writer` instead of `w`, since that name is used in all other methods of this class.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java [hunk 8]
```diff
@@ -296,4 +296,5 @@ private boolean hasMemberLeft(Invocation invocation) {
             }
         }
     }
+
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java]: Useless new empty line? :D

File: hazelcast/src/test/java/com/hazelcast/Main.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+import com.hazelcast.map.EntryBackupProcessor;
+import com.hazelcast.map.EntryProcessor;
+import com.hazelcast.test.HazelcastTestSupport;
+
+import java.util.Map;
+
+/**
+ * Created by alarmnummer on 1/23/16.
+ */
+public class Main extends HazelcastTestSupport {
```
Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/Main.java]: Should this class stay in the codebase? Looks like a WIP for a new test.

File: hazelcast/src/test/java/com/hazelcast/internal/monitors/OverloadedConnectionsPluginTest.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+import static com.hazelcast.instance.GroupProperty.PERFORMANCE_MONITOR_OVERLOADED_CONNECTIONS_PERIOD_SECONDS;
+import static com.hazelcast.instance.GroupProperty.PERFORMANCE_MONITOR_OVERLOADED_CONNECTIONS_SAMPLES;
+import static com.hazelcast.instance.GroupProperty.PERFORMANCE_MONITOR_OVERLOADED_CONNECTIONS_THRESHOLD;
+
+/**
+ * This test can't run with test-hazelcast instances since we rely on a real TcpIpConnectionManager.
+ */
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(QuickTest.class)
+@Ignore
```
Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/internal/monitors/OverloadedConnectionsPluginTest.java]: Is the test intentionally ignored (can't tell by just looking at it, if it should work already)?

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperty.java [hunk 11]
```diff
@@ -148,6 +141,21 @@
      */
     PERFORMANCE_MONITOR_HUMAN_FRIENDLY_FORMAT("hazelcast.performance.monitor.human.friendly.format", true),
 
+    PERFORMANCE_MONITOR_SLOW_OPERATIONS_PERIOD_SECONDS("hazelcast.performance.monitor.slowoperations.period.seconds", 60),
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperty.java]: You should define these properties with a `TimeUnit`, so you can benefit from the `getMillis()`, `getSeconds()` methods in `HazelcastProperties`, e.g. `PERFORMANCE_MONITOR_SLOW_OPERATIONS_PERIOD_SECONDS("hazelcast.performance.monitor.slowoperations.period.seconds", 60, SECONDS),`

File: hazelcast/src/test/java/com/hazelcast/Main.java (hunk 9 above)
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/Main.java]: Will be deleted. I'm using it to check the content of the generated output.

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/SlowOperationPlugin.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    public SlowOperationPlugin(NodeEngineImpl nodeEngine) {
+        this.logger = nodeEngine.getLogger(SlowOperationPlugin.class);
+        this.operationService = nodeEngine.getOperationService();
+        this.periodSeconds = getPeriodSeconds(nodeEngine);
+    }
+
+    private long getPeriodSeconds(NodeEngineImpl nodeEngine) {
+        GroupProperties groupProperties = nodeEngine.getGroupProperties();
+        if (!groupProperties.getBoolean(SLOW_OPERATION_DETECTOR_ENABLED)) {
+            return -1;
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/SlowOperationPlugin.java]: I think the disabled magic number was `0`, not `-1`. Maybe another good candidate for a constant.

**Comment by mdogan** [comment]
As @jerrinot said, having more data about invocations would be nice, like; 
- how many local and remote
- how many waiting for actual response
- how many have done more than X retries
- how many waiting for backup response
- how many pending per target (?)

**Comment by mdogan** [comment]
Can we also monitor heartbeat timestamps? When each member did send last heartbeat...

**Comment by pveentjer** [comment]
@mdogan I believe that a lot of your nice to haves can be implemented (if they are not already) as metrics. There is a metrics performancemonitor plugin that dumps everything in the metrics registry.

I'll check them and if missing, I'll add missing probes.

The backup stuff I'll add; good to have. 
Pending per target ----> good to have.

File: hazelcast/src/main/java/com/hazelcast/internal/metrics/renderers/CommaSeparatedKeyValueProbeRenderer.java [hunk 13]
```diff
@@ -32,7 +32,7 @@
     @Override
     public void start() {
         sb.setLength(0);
-        sb.append("{").append("time").append("=").append(System.currentTimeMillis());
+        sb.append("").append("time=").append(System.currentTimeMillis());
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/metrics/renderers/CommaSeparatedKeyValueProbeRenderer.java]: Line still contains an empty `.append("")`.

File: checkstyle/suppressions.xml [hunk 14]
```diff
@@ -272,6 +272,9 @@
     <suppress checks="MethodCount|NPathComplexity|MagicNumber" files="com/hazelcast/spi/Operation"/>
     <suppress checks="MethodCount" files="com/hazelcast/spi/impl/NodeEngineImpl"/>
     <suppress checks="ClassFanOutComplexity" files="com/hazelcast/spi/impl/NodeEngineImpl"/>
+    <!-- node engine constructs objects, so it knows many of them -->
+    <suppress checks="ClassDataAbstractionCoupling" files="com/hazelcast/spi/impl/NodeEngineImpl"/>
```
Comment (mtopolnik) [inline checkstyle/suppressions.xml]: I strongly suggest migrating away from suppressions.xml towards `@SuppressWarnings("checkstyle:classdataabstractioncoupling")`. It makes it much more visible that something was suppressed, plus it discourages wildcard suppressions.

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/PerformanceMonitor.java [hunk 15]
```diff
@@ -16,111 +16,135 @@
...
+        checkNotNull(performanceMonitorPlugin, "monitorTask can't be null");
+
         if (!enabled) {
-            return null;
+            return;
+        }
+
+        long periodMillis = performanceMonitorPlugin.getPeriodMillis();
+
+        logger.finest(performanceMonitorPlugin.getClass().toString() + " is " + (periodMillis == 0 ? "disabled" : "enabled"));
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/PerformanceMonitor.java]: You could use the `DISABLED` in here as well.

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/PerformanceMonitor.java [hunk 16]
```diff
@@ -16,111 +16,135 @@
...
+
+        if (periodMillis == 0) {
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/PerformanceMonitor.java]: Use case for `DISABLED` as well.

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/PerformanceMonitor.java [hunk 17]
```diff
@@ -16,111 +16,135 @@
...
-        HazelcastThreadGroup threadGroup = node.getHazelcastThreadGroup();
-        int delaySeconds = node.getGroupProperties().getSeconds(GroupProperty.PERFORMANCE_MONITOR_DELAY_SECONDS);
+        performanceMonitorPlugin.setSingleLine(singleLine);
+        performanceMonitorPlugin.onStart();
 
-        return new MonitorThread(threadGroup, delaySeconds);
+        if (periodMillis > 0) {
+            // it is a periodic task
+            scheduler.scheduleAtFixedRate(new MonitorTaskRunnable(performanceMonitorPlugin), 0, periodMillis, MILLISECONDS);
+        } else {
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/PerformanceMonitor.java]: This looks nice and clean, but in fact every negative `periodMillis` will execute the plugin once, although it's documented to be `-1` (aka `RUN_ONCE`). Maybe we should check with `else if (periodMillis == RUN_ONCE)` here and throw an `IllegalArgumentException` for all other negative values? Don't know what our policy for wrong configurations is (if there is any).

File: hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java [hunk 18]
```diff
@@ -138,6 +148,18 @@ public void start() {
         proxyService.init();
         operationService.start();
         quorumService.start();
+        performanceMonitor.start();
+
+        // static loggers at beginning of file
+        performanceMonitor.register(new BuildInfoPlugin(this));
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java]: Maybe we can do an automatic registration via reflection and annotations? With this solution you have to know about this part of code (and NodeEngineImpl has to know about all plugins). We could also get rid of the `ClassDataAbstractionCoupling` suppression. Could be an optimization for another PR.

File: checkstyle/suppressions.xml (hunk 14 above)
Comment (Donnerbart) [inline checkstyle/suppressions.xml]: I'm in favor of the annotation style as well. Another advantage: If you refactor a class you directly see the suppressions and can give it a try to remove them. A suppression in the XML is easily forgotten, even if the class is improved.

But I won't make this a blocker if we can't get a consensus on this. Would be an easy task to convert most of the XML config to annoations in a separate PR.

I also made a suggestion to get rid of this specific suppression for `NodeEngineImpl` (no suppression at all should be the best solution anyway).

**Comment by Donnerbart** [comment]
I like the general design of the new PerformanceMonitor. The plugins are a very good idea to break down this complex task.

There are some minor remarks here and there, but I see no big blocker for an initial version. Missing metrics can be added anytime later, as well as some optimiztations in some of the plugins.

File: checkstyle/suppressions.xml (hunk 14 above)
Comment (mtopolnik) [inline checkstyle/suppressions.xml]: Another case in point: I already had a PR prepared which removes the blanket XML suppression from the complete `c.h.elastic` hierarchy. The PR was delayed until after the release; but in the meantime another PR was accepted which turned out to add new code to the same package hierarchy, even though in a completely separate module. After rebasing I was welcomed with 300 new CheckStyle errors to take care of.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java (hunk 18 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java]: THe NodeEngineImpl is acting like a application context; so the application gets constructed here. That is why it is perfectly fine that the NodeEngineImpl is violating this class data abstraction coupling.

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/MetricsPlugin.java [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        @Override
+        public void renderDouble(String name, double value) {
+            writer.writeKeyValueEntry(name, value);
+        }
+
+        @Override
+        public void renderException(String name, Exception e) {
+            writer.writeKeyValueEntry(name, e.getClass().getName() + ":" + e.getMessage());
```
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/MetricsPlugin.java]: `":"` can be `':'`

**Comment by mtopolnik** [comment]
:+1:

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperty.java [hunk 20]
```diff
@@ -114,14 +114,7 @@
      */
     PERFORMANCE_METRICS_LEVEL("hazelcast.performance.metric.level", ProbeLevel.MANDATORY.name()),
 
-    /**
-     * The delay in seconds between monitor of the performance.
-     * <p/>
-     * The default is 30 seconds.
-     */
-    PERFORMANCE_MONITOR_DELAY_SECONDS("hazelcast.performance.monitor.delay.seconds", 30, SECONDS),
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperty.java]: should we deprecate it / fail-fast? now the property will be simple ignored -> not nice.

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/MultiLinePerformanceLogWriter.java [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+ * A {@link PerformanceLogWriter} that writes over multiple lines. Useful for human reading.
+ */
+public class MultiLinePerformanceLogWriter extends PerformanceLogWriter {
+
+    private static final String STR_LONG_MIN_VALUE = String.format("%,d", Long.MIN_VALUE);
+
+    private static final char[] DIGITS = new char[]{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9'};
+
+    private static final String[] INDENTS = new String[]{
+            "\n                          ",
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/MultiLinePerformanceLogWriter.java]: platform-specific stuff

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/OverloadedConnectionsPlugin.java [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+ *
+ * Currently the sampling has a lot of overhead since the value needs to be deserialized. That is why this
+ * plugin is disabled by default.
+ */
+public class OverloadedConnectionsPlugin extends PerformanceMonitorPlugin {
+    private static final Queue<OutboundFrame> EMPTY_QUEUE = new LinkedList<OutboundFrame>();
+
+    private final SerializationService serializationService;
+    private final ILogger logger;
+    private final Map<String, MutableInteger> occurrenceMap = new HashMap<String, MutableInteger>();
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/OverloadedConnectionsPlugin.java]: a small suggestion: `ItemCounter` might be a more suitable abstraction for this. It miss iterations, but it would be easier to add.

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperty.java (hunk 20 above)
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperty.java]: Hmm, maybe we can create a section where we can put outdated properties and some kind of help text (reference to new property, explanation etc.). With this we could keep the existing properties clean and have a central place for the historical stuff.

**Comment by jerrinot** [comment]
:+1:  once the platform specific end of lines are removed.

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/MultiLinePerformanceLogWriter.java (hunk 21 above)
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/MultiLinePerformanceLogWriter.java]: In Simulator I put this in our `FormatUtils`, so I can use it in the rest of the code:
`public static final String NEW_LINE = System.getProperty("line.separator");`

In Hazelcast it seems to capsulated in this method: `StringUtil.getLineSeperator()`

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/MultiLinePerformanceLogWriter.java]: I don't think it is platform specific.

1: any decent editor on windows or osx will deal with /n correctly
2: this stuff is typically run on linux; but viewed on a range of platforms like windows/linux/osx. So you already got a difference between your running platform and your viewing platform.

Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/MultiLinePerformanceLogWriter.java]: fair enough. my opinion is not too strong here.

**Comment by Donnerbart** [comment]
:+1:

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/OverloadedConnectionsPlugin.java (hunk 22 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/OverloadedConnectionsPlugin.java]: Can you elaborate?

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperty.java (hunk 20 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperty.java]: These should not be officially supported properties. They are there for debugging only

File: hazelcast/src/main/java/com/hazelcast/internal/monitors/MultiLinePerformanceLogWriter.java (hunk 21 above)
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/internal/monitors/MultiLinePerformanceLogWriter.java]: A sidenote: writing `String.format("%n                              ")` takes care of the line separator, if you ever feel the need for that.
